overlord/snapshotstate/snapshotstate.go

```diff
@@ -0,0 +1,264 @@
+			Users: users,
+		}
+		task.Set("shot", &shot)
+		task.Set("snap-setup", &snapstate.SnapSetup{SideInfo: &snap.SideInfo{RealName: name}})
+		ts.AddTask(task)
+	}
+
+	return shID, snapNames, ts, nil
+}
+
+// Restore creates a taskset for restoring a snapshot's data.
+// Note that the state must be locked by the caller.
+func Restore(st *state.State, snapshotID uint64, snapNames []string, users []string) ([]string, *state.TaskSet, error) {
+	snapNames, filenames, err := snapNamesInSnapshot(snapshotID, snapNames)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := snapstate.CheckChangeConflictMany(st, snapNames, nil); err != nil {
+		return nil, nil, err
+	}
+
+	// restore needs to conflict with forget of itself
+	if err := checkSnapshotChangeConflict(st, snapshotID, "forget-snapshot"); err != nil {
+		return nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for i, name := range snapNames {
+		desc := fmt.Sprintf("Restoring data from snapshot #%d for snap %q", snapshotID, name)
+		task := st.NewTask("restore-snapshot", desc)
+		shot := shotState{
+			ID:       snapshotID,
+			Snap:     name,
+			Users:    users,
+			Filename: filenames[i],
+		}
+		task.Set("shot", &shot)
+		// hackish, for conflict detection:
+		task.Set("snap-setup", &snapstate.SnapSetup{SideInfo: &snap.SideInfo{RealName: name}})
+		ts.AddTask(task)
+	}
+
+	return snapNames, ts, nil
+}
+
+// Check creates a taskset for checking a snapshot's data.
+// Note that the state must be locked by the caller.
+func Check(st *state.State, snapshotID uint64, snapNames []string, users []string) ([]string, *state.TaskSet, error) {
+	// check needs to conflict with forget of itself
+	if err := checkSnapshotChangeConflict(st, snapshotID, "forget-snapshot"); err != nil {
+		return nil, nil, err
+	}
+
+	snapNames, filenames, err := snapNamesInSnapshot(snapshotID, snapNames)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for i, name := range snapNames {
+		desc := fmt.Sprintf("Checking data in snapshot #%d for snap %q", snapshotID, name)
+		task := st.NewTask("check-snapshot", desc)
+		shot := shotState{
+			ID:       snapshotID,
+			Snap:     name,
+			Users:    users,
+			Filename: filenames[i],
+		}
+		task.Set("shot", &shot)
+		ts.AddTask(task)
+	}
+
+	return snapNames, ts, nil
+}
+
+// Forget creates a taskset for deletinig a snapshot.
+// Note that the state must be locked by the caller.
+func Forget(st *state.State, snapshotID uint64, snapNames []string) ([]string, *state.TaskSet, error) {
+	// forget needs to conflict with check and restore
+	if err := checkSnapshotChangeConflict(st, snapshotID, "check-snapshot", "restore-snapshot"); err != nil {
+		return nil, nil, err
+	}
+
+	snapNames, filenames, err := snapNamesInSnapshot(snapshotID, snapNames)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for i, name := range snapNames {
+		desc := fmt.Sprintf("Removing snapshot #%d for snap %q", snapshotID, name)
```

Comment:
"Remove snapshot set #%d for snap %q", but that sounds reversed. This should probably be something like "Remove data of snap %q from snapshot set #%d".